skfp/utils/splitters.py

```diff
@@ -0,0 +1,239 @@
+    Journal of Medicinal Chemistry, 39(15), 2887-2893.
+    .. [2] Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande (2017).
+    MoleculeNet: A Benchmark for Molecular Machine Learning. Chemical Science, 9(2), 513-530.
+    """
+    np.testing.assert_almost_equal(train_size + test_size + valid_size, 1.0)
+
+    scaffolds = _create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_cutoff: int = int(train_size * len(data))
+    valid_cutoff: int = int((train_size + valid_size) * len(data))
+
+    train_ids: list[int] = []
+    valid_ids: list[int] = []
+    test_ids: list[int] = []
+
+    for scaffold_set in scaffold_sets:
+        if len(train_ids) < train_cutoff:
+            train_ids.extend(scaffold_set)
+        elif len(valid_ids) < valid_cutoff:
+            valid_ids.extend(scaffold_set)
+        else:
+            test_ids.extend(scaffold_set)
+
+    if return_indices:
+        return train_ids, valid_ids, test_ids
+    else:
+        return (
+            _get_data_from_indices(data, scaffolds, train_ids),
+            _get_data_from_indices(data, scaffolds, valid_ids),
+            _get_data_from_indices(data, scaffolds, test_ids),
+        )
+
+
+def _create_scaffolds(
+    data: list[Union[str, Mol]], include_chirality: bool = False
+) -> dict[str, list]:
+    """
+    Generate Bemis-Murcko scaffolds for a list of SMILES strings or RDKit 'Mol' objects.
+
+    Parameters
+    ----------
+    data : sequence
+        List of SMILES strings or RDKit 'Mol' objects.
+
+    include_chirality: bool, default=False
+        Whether to take chirality of molecules into consideration.
+
+    Returns
+    -------
+    scaffolds : dict
+        A dictionary where keys are Bemis-Murcko scaffolds (as SMILES strings) and
+        values are lists of indices pointing to molecules sharing the same scaffold.
+
+    Notes
+    -----
+    - This implementation uses Bemis-Murcko scaffolds (Bemis and Murcko, 1996) to group molecules.
+    - Each scaffold is represented as a SMILES string.
+    """
+    scaffolds = defaultdict(list)
+    molecules = ensure_mols(data)
+
+    for ind, molecule in enumerate(molecules):
+        scaffold = MurckoScaffold.MurckoScaffoldSmiles(
+            mol=molecule, includeChirality=include_chirality
+        )
+        scaffolds[scaffold].append(ind)
+
+    return scaffolds
+
+
+def _get_data_from_indices(
+    data: Sequence[Union[str, Mol]], scaffolds: dict[str, list], indices: Sequence[int]
+) -> list[Union[str, Mol]]:
+    """
+    Helper function to retrieve data from indices.
+    """
+    result = []
+    index_set = set(indices)
+
+    for i in index_set:
+        if isinstance(data[i], str):
+            item = data[i]
+        elif isinstance(data[i], Mol):
+            item = data[i]
+        else:
+            raise ValueError("Unsupported data type in input.")
```

Comment:
The type doesn't really matter here, you are just taking the indexes. This can be simplified to:
```
indices = set(indices)
return [data[idx] for idx in indices]
```